experimental/lib/Transforms/ResourceSharing/Crush.cpp

```diff
@@ -0,0 +1,682 @@
+            sharedOp->getLoc(), sharingWrapperOutputTypes, sharingWrapperInputs,
+            namedCreditsAttr);
+
+    for (auto [id, op] : llvm::enumerate(group)) {
+      op->getResult(0).replaceAllUsesWith(wrapperOp->getResult(id));
+    }
+
+    for (auto [id, val] : llvm::enumerate(sharedOp->getOperands())) {
+      sharedOp->replaceUsesOfWith(val, wrapperOp->getResult(id + group.size()));
+    }
+
+    // Remove all operations in the sharing group except for the shared one.
+    for (auto *op : group) {
+      if (op != *(group.begin())) {
+        op->erase();
+      }
+    }
+  }
+
+  return success();
+}
+
+void CreditBasedSharingPass::runDynamaticPass() {
+  MLIRContext *ctx = &getContext();
+  OpBuilder builder(ctx);
+  NameAnalysis &namer = getAnalysis<NameAnalysis>();
+
+  TimingDatabase timingDB(&getContext());
+  if (failed(TimingDatabase::readFromJSON(timingModels, timingDB)))
+    signalPassFailure();
+
+  // Buffer placement requires that all values are used exactly once
+  ModuleOp modOp = getOperation();
+  if (failed(verifyIRMaterialized(modOp))) {
+    modOp->emitError() << ERR_NON_MATERIALIZED_MOD;
+    return;
+  }
+
+  SharingInfo sharingInfo;
+
+  // Run buffer placement pass and fill sharingInfo with performance analysis
+  // information
+  if (failed(runBufferPlacementPass(modOp, sharingInfo))) {
+    modOp->emitError() << "Failed running buffer placement";
+    return;
+  }
+
+  // If buffers are placed naively, then no critical CFC is set for each funcOp.
+  // We can also share operations naively.
+  if (algorithm == ON_MERGES) {
+    for (handshake::FuncOp funcOp :
+         getOperation().getOps<handshake::FuncOp>()) {
+      FuncPerfInfo funcPerfInfo;
+      sharingInfo[&funcOp] = funcPerfInfo;
+    }
+  }
+
+  for (auto &[funcOp, funcPerfInfo] : sharingInfo) {
+
+    // Check the sharing targets
+    SmallVector<Operation *> sharingTargets = getSharingTargets(*funcOp);
+
+    // opOccupancy: maps each operation to the maximum occupancy it has to
+    // achieve.
+    llvm::MapVector<Operation *, double> opOccupancy;
+    getOpOccupancy(sharingTargets, opOccupancy, timingDB, funcPerfInfo);
+
+    // Initialize the sharing groups:
+    SharingGroups sharingGroups;
+    for (auto [id, op] : llvm::enumerate(sharingTargets)) {
+      Group g;
+      g.push_back(op);
+      sharingGroups.push_back(g);
+    }
+
+    // Determine SCCs
+    for (auto critCfc : funcPerfInfo.critCfcs) {
+      std::map<Operation *, size_t> sccMap = getSccsInCfc(
+          funcPerfInfo.cfUnits[critCfc], funcPerfInfo.cfChannels[critCfc]);
+      funcPerfInfo.cfSccs.emplace(critCfc, sccMap);
+    }
+
+    llvm::errs() << "Initial groups\n";
+    logGroups(sharingGroups, namer);
+
+    // Merge groups
+    for (bool continueMerging = true; continueMerging;) {
+      continueMerging = tryMergeGroups(sharingGroups, funcPerfInfo);
+    }
+
+    // log
+    llvm::errs() << "Finished merging\n";
```

Comment:
Nothing should be printed to `llvm::errs()/stderr` unless it is actually an error (and without explicit debug information requested nothing should be printed on `llvm::outs()/stdout` either).